vars/buildPlugin.groovy

```diff
@@ -61,7 +61,14 @@ def call(Map params = [:]) {
     }
 
     tasks[stageIdentifier] = {
+      int retryCounter = 0
       retry(count: 3, conditions: [kubernetesAgent(handleNonKubernetes: true), nonresumable()]) {
+        if (retryCounter > 1 && platform != 'windows') {
+          // no spot instances for windows for now TODO change when available
+          // Use a spot instance for the 2 first try and nonspot for third and last
+          label = 'ubuntu-22-amd64-maven17-nonspot'
```

Comment:
Why not just use labels the way they were intended?
```suggestion
        if (retryCounter > 1) {
          // Use a spot instance (if applicable) for the 2 first try and nonspot for third and last
          label += ' && !spot'
```
assuming of course that you have marked the usual Linux agents with the `spot` label. Then there is no need to hard-code a particular label expression for the rest, and if we do add Windows agents supporting Spot later, fine: just mark them with the `spot` label, and offer an alternate pool without it for on-demand.

Reply:
if we just add a `spot` label and use your idea of && spot or && !spot it's working for the shared pipeline library but we miss the point of other usage, meaning we want almost all call to agent to fall on spot ones, so default is spot ... and exception (or retry) is non spot ...

if both spot and nonspot got the same set of label, lets say vm && linux ... both can be called ...

if I understood correctly ...

Reply:
Yeah you would want to use `&& spot` for the first two tries. Are there that many other pieces of Pipeline script running the system? `bom/Jenkinsfile`, `jenkins/Jenkinsfile`, and a few miscellaneous low-traffic projects I guess?

Reply:
(You may be able to designate Spot-only `Cloud`s as first in the list and hope they service requests preferentially. Not sure.)

Reply:
> (You may be able to designate Spot-only Clouds as first in the list and hope they service requests preferentially. Not sure.)

With EC2 plugins, for sure it does not work :'( (it works with ACI, kubernetes and auzr-evm though)

Reply:
see https://github.com/jenkins-infra/jenkins-infra/pull/3978 that set the labels automatically and this one have been updated to match spot and then nonspot agents